File: javascript/lesson 4/quickSort.js
// https://leetcode.com/problems/sort-an-array/

function sortArray(nums) {
    quickSort(nums);
    return nums;
}

function quickSort(nums) {
    shuffle(nums);
    sort(nums, 0, nums.length - 1);
}

function sort(nums, lo, hi) {
    if (lo >= hi) return;
    const pivot = partition(nums, lo, hi);
    sort(nums, lo, pivot - 1);
    sort(nums, pivot + 1, hi);
}

function partition(nums, lo, hi) {
    const pivot = nums[lo];
    let i = lo;
    let j = hi + 1;
    while (true) {
        while (nums[++i] < pivot && i < hi);
        while (nums[--j] > pivot && j > lo);
        if (i >= j) break;
        swap(nums, i, j);
    }
    swap(nums, lo, j);
    return j;
}

function swap(nums, a, b) {
    [nums[a], nums[b]] = [nums[b], nums[a]];
}

function shuffle(nums) {
    for (let i = 0; i < nums.length; i++) {
        const r = Math.floor(Math.random() * (i + 1));
        swap(nums, i, r);
    }    
}

File: python/lesson 4/quick_sort.py
import random
from typing import List

# https://leetcode.com/problems/sort-an-array/
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        self.quick_sort(nums)
        return nums
    
    def quick_sort(self, nums: List[int]):
        self.shuffle(nums)
        self.sort(nums, 0, len(nums) - 1)
    
    def sort(self, nums: List[int], lo: int, hi: int):
        if hi <= lo: return
        pivot = self.partition(nums, lo, hi)
        self.sort(nums, lo, pivot - 1)
        self.sort(nums, pivot + 1, hi)
    
    def partition(self, nums: List[int], lo: int, hi: int):
        pivot = nums[lo]
        i = lo
        j = hi + 1
        while True:
            while True:
                i += 1
                if not (nums[i] < pivot and i < hi): break
            while True:
                j -= 1
                if not (nums[j] > pivot and j > lo): break            
            if i >= j: break
            self.swap(nums, i, j)
        self.swap(nums, lo, j)
        return j
    
    def swap(self, nums: List[int], a: int, b: int):
        nums[a], nums[b] = nums[b], nums[a]
        
        
    def shuffle(self, nums):
        for i in range(len(nums)):
            r = random.randrange(0, i + 1)        
            self.swap(nums, i, r)
